karabo_data/geometry2.py

```diff
@@ -0,0 +1,483 @@
+                rotn = np.stack((ss_vec_yx, fs_vec_yx), axis=-1)
+
+                # affine_transform takes a mapping from *output* to *input*.
+                # So we reverse the forward transformation.
+                transform = np.linalg.inv(rotn)
+                offset = np.dot(rotn, corner_pos_yx)  # this seems to work, but is it right?
+
+                affine_transform(tile_data, transform, offset=offset, cval=np.nan,
+                                 output_shape=size_yx, output=tmp[i * 8 + j])
+
+        # Silence warnings about nans - we expect gaps in the result
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=RuntimeWarning)
+            out = np.nanmax(tmp, axis=0)
+
+        return out, centre
+
+    def _plotting_dimensions(self):
+        """Calculate appropriate dimensions for plotting assembled data
+
+        Returns (size_x, size_y), (centre_x, centre_y)
+        """
+        corners = []
+        for module in self.modules:
+            for tile in module:
+                corners.append(tile.corners())
+        corners = np.concatenate(corners)[:, :2]
+
+        # Find extremes, add 20 px margin
+        min_xy = corners.min(axis=0).astype(int) - 20
+        max_xy = corners.max(axis=0).astype(int) + 20
+
+        size = max_xy - min_xy
+        centre = -min_xy
+        return tuple(size), centre
+
+    def plot_data(self, modules_data):
+        """Plot data from the detector using this geometry.
+
+        Returns a matplotlib figure.
+
+        Parameters
+        ----------
+
+        modules_data : ndarray
+          Should have exactly 3 dimensions: channelno, pixel_ss, pixel_fs
+          (lengths 16, 512, 128). ss/fs are slow-scan and fast-scan.
+        """
+        from matplotlib.cm import viridis
+        from matplotlib.backends.backend_agg import FigureCanvasAgg
+        from matplotlib.figure import Figure
+
+        fig = Figure((10, 10))
+        FigureCanvasAgg(fig)
+        ax = fig.add_subplot(1, 1, 1)
+        my_viridis = copy(viridis)
+        # Use a dark grey for missing data
+        my_viridis.set_bad('0.25', 1.)
+
+        res, centre = self.position_all_modules(modules_data)
+        ax.imshow(res, origin='lower', cmap=my_viridis)
+
+        cx, cy = centre
+        ax.hlines(cy, cx - 20, cx + 20, colors='w', linewidths=1)
+        ax.vlines(cx, cy - 20, cy + 20, colors='w', linewidths=1)
+        return fig
+
+    def snap(self):
+        """Snap geometry to a 2D pixel grid
+
+        This returns a new geometry object. The 'snapped' geometry is
+        less accurate, but can assemble data into a 2D array more efficiently,
+        because it doesn't do any interpolation.
+        """
+        new_modules = []
+        for module in self.modules:
+            new_tiles = [t.snap() for t in module]
+            new_modules.append(new_tiles)
+        return AGIPD_1M_SnappedGeometry(new_modules)
+
+class AGIPD_1M_SnappedGeometry:
```

Comment:
What about making this the default class? So that you don't have to call the snap method to access the snapped geometry?

Reply:
Yes that sounds like a good idea! Let's do that then